components/web_discovery/browser/server_config_loader.cc

```diff
@@ -0,0 +1,446 @@
+  collector_config_url_ =
+      GURL(GetAnonymousHPNHost() + kCollectorConfigPathWithFields);
+  quorum_config_url_ = GURL(GetQuorumHost() + kQuorumConfigPath);
+  patterns_url_ = GetPatternsEndpoint();
+
+  patterns_path_ = user_data_dir.AppendASCII(kPatternsFilename);
+}
+
+ServerConfigLoader::~ServerConfigLoader() = default;
+
+const ServerConfig& ServerConfigLoader::GetLastServerConfig() const {
+  CHECK(last_loaded_server_config_);
+  return *last_loaded_server_config_;
+}
+
+const PatternsGroup& ServerConfigLoader::GetLastPatterns() const {
+  CHECK(last_loaded_patterns_);
+  return *last_loaded_patterns_;
+}
+
+void ServerConfigLoader::SetLastServerConfigForTesting(
+    std::unique_ptr<ServerConfig> server_config) {
+  last_loaded_server_config_ = std::move(server_config);
+}
+
+void ServerConfigLoader::SetLastPatternsForTesting(
+    std::unique_ptr<PatternsGroup> patterns) {
+  last_loaded_patterns_ = std::move(patterns);
+}
+
+void ServerConfigLoader::LoadConfigs() {
+  if (collector_config_url_loader_ || quorum_config_url_loader_) {
+    // Another request is in progress
+    return;
+  }
+  auto collector_resource_request =
+      CreateResourceRequest(collector_config_url_);
+  auto quorum_resource_request = CreateResourceRequest(quorum_config_url_);
+
+  collector_config_url_loader_ = network::SimpleURLLoader::Create(
+      std::move(collector_resource_request), kNetworkTrafficAnnotation);
+  quorum_config_url_loader_ = network::SimpleURLLoader::Create(
+      std::move(quorum_resource_request), kNetworkTrafficAnnotation);
+
+  auto callback = base::BarrierCallback<std::optional<std::string>>(
+      2, base::BindOnce(&ServerConfigLoader::OnConfigResponses,
+                        base::Unretained(this)));
+
+  collector_config_url_loader_->DownloadToString(
+      shared_url_loader_factory_.get(), callback, kMaxResponseSize);
+  quorum_config_url_loader_->DownloadToString(shared_url_loader_factory_.get(),
+                                              callback, kMaxResponseSize);
+}
+
+void ServerConfigLoader::OnConfigResponses(
+    std::vector<std::optional<std::string>> response_bodies) {
+  CHECK_EQ(response_bodies.size(), 2u);
+  base::Time update_time = base::Time::Now();
+  bool result = ProcessConfigResponses(response_bodies[0], response_bodies[1]);
+
+  config_backoff_entry_.InformOfRequest(result);
+
+  collector_config_url_loader_ = nullptr;
+  quorum_config_url_loader_ = nullptr;
+
+  if (!result) {
+    update_time += config_backoff_entry_.GetTimeUntilRelease();
+  } else {
+    update_time += base::RandTimeDelta(kMinReloadInterval, kMaxReloadInterval);
+
+    SchedulePatternsRequest();
+  }
+
+  config_update_timer_.Start(
+      FROM_HERE, update_time,
+      base::BindOnce(&ServerConfigLoader::LoadConfigs, base::Unretained(this)));
+}
+
+bool ServerConfigLoader::ProcessConfigResponses(
+    const std::optional<std::string>& collector_response_body,
+    const std::optional<std::string>& quorum_response_body) {
```

Comment:
I don't think this is correct. The doc says: "The ordering of the vector is unspecified".